Author: Luca Boccassi <luca.boccassi@gmail.com>
Description: Fix kernel module build on 4.10 and greater
 From kernel 4.10 and newer (commit 530e9b76ae8f8) CPU_DOWN_FAILED and
 CPU_DOWN_PREPARE are no longer available as events, together with their
 callback registers.
--- a/nv-pat.c
+++ b/nv-pat.c
@@ -210,19 +210,23 @@
 
     switch (action)
     {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 10, 0)
         case CPU_DOWN_FAILED:
+#endif
         case CPU_ONLINE:
             if (cpu == (NvUPtr)hcpu)
                 nv_setup_pat_entries(NULL);
             else
                 NV_SMP_CALL_FUNCTION(nv_setup_pat_entries, hcpu, 1);
             break;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 10, 0)
         case CPU_DOWN_PREPARE:
             if (cpu == (NvUPtr)hcpu)
                 nv_restore_pat_entries(NULL);
             else
                 NV_SMP_CALL_FUNCTION(nv_restore_pat_entries, hcpu, 1);
             break;
+#endif
     }
 
     put_cpu();
@@ -252,7 +256,7 @@
     if (!disable_pat)
     {
         nv_enable_pat_support();
-#if defined(NV_ENABLE_PAT_SUPPORT) && defined(NV_ENABLE_HOTPLUG_CPU)
+#if defined(NV_ENABLE_PAT_SUPPORT) && defined(NV_ENABLE_HOTPLUG_CPU) && LINUX_VERSION_CODE < KERNEL_VERSION(4, 10, 0)
         if (nv_pat_mode == NV_PAT_MODE_BUILTIN)
         {
             if (register_hotcpu_notifier(&nv_hotcpu_nfb) != 0)
@@ -279,7 +283,7 @@
     if (nv_pat_mode == NV_PAT_MODE_BUILTIN)
     {
         nv_disable_pat_support();
-#if defined(NV_ENABLE_PAT_SUPPORT) && defined(NV_ENABLE_HOTPLUG_CPU)
+#if defined(NV_ENABLE_PAT_SUPPORT) && defined(NV_ENABLE_HOTPLUG_CPU) && LINUX_VERSION_CODE < KERNEL_VERSION(4, 10, 0)
         unregister_hotcpu_notifier(&nv_hotcpu_nfb);
 #endif
     }
